qiskit_optimization/algorithms/warm_start_qaoa_optimizer.py

```diff
@@ -0,0 +1,339 @@
+        """
+        Creates an initial state quantum circuit to warm start QAOA.
+
+        Args:
+            initial_variables: Already created initial variables.
+
+        Returns:
+            A quantum circuit that represents initial state.
+        """
+        circuit = QuantumCircuit(len(initial_variables))
+
+        for index, relaxed_value in enumerate(initial_variables):
+            theta = 2 * np.arcsin(np.sqrt(relaxed_value))
+            circuit.ry(theta, index)
+
+        return circuit
+
+    def create_mixer(self, initial_variables: List[float]) -> QuantumCircuit:
+        """
+        Creates an evolved mixer circuit as Ry(theta)Rz(-2beta)Ry(-theta).
+
+        Args:
+            initial_variables: Already created initial variables.
+
+        Returns:
+            A quantum circuit to be used as a mixer in QAOA.
+        """
+        circuit = QuantumCircuit(len(initial_variables))
+        beta = Parameter("beta")
+
+        for index, relaxed_value in enumerate(initial_variables):
+            theta = 2 * np.arcsin(np.sqrt(relaxed_value))
+
+            circuit.ry(-theta, index)
+            circuit.rz(-2.0 * beta, index)
+            circuit.ry(theta, index)
+
+        return circuit
+
+
+class WarmStartQAOAOptimizer(MinimumEigenOptimizer):
+    # todo: update documentation
+    """A meta-algorithm that uses a pre-solver to solve a relaxed version of the problem.
+    Users must implement their own pre solvers by inheriting from the base class."""
+
+    def __init__(self,
+                 pre_solver: OptimizationAlgorithm,
+                 relax_for_pre_solver: bool,
+                 qaoa: QAOA,
+                 epsilon: float = 0.25,
+                 num_initial_solutions: int = 1,
+                 circuit_factory: Optional[WarmStartQAOACircuitFactory] = None,
+                 aggregator: Optional[BaseAggregator] = None,
+                 penalty: Optional[float] = None,
+                 converters: Optional[Union[QuadraticProgramConverter,
+                                            List[QuadraticProgramConverter]]] = None
+                 ) -> None:
+        """ Initializes the optimizer. For correct initialization either
+            ``epsilon`` or ``circuit_factory`` must be passed. If only ``epsilon`` is specified
+            (either an explicit value or the default one), then an instance of
+            :class:`~qiskit.optimization.algorithms.WarmStartQAOACircuitFactory` is created.
+            If ``circuit_factory`` is specified then this instance is used in the implementation
+            and ``epsilon`` value is ignored.
+
+        Args:
+            pre_solver: An instance of an optimizer to solve the relaxed version of the problem.
+            relax_for_pre_solver: True if the problem must be relaxed to the continuous case
+                before passing it to the pre-solver.
+            qaoa: A QAOA instance to be used in the computations.
```

Comment:
Very specifically we need an instance of the QAOA algorithm from Terra right since it will be updated and we expect it to have certain setters - at least if anyone created their own QAOA it would have to conform in that regard

Reply:
Yes, we expect the one from Terra.